src/Common/src/System/StringNormalizationExtensions.Unix.cs

```diff
@@ -6,35 +6,120 @@
 
 namespace System
 {
-    // TODO: We need an actual implementation here, but since all ASCII strings are normalized under all normalization forms
-    // we can handle the simple cases, which will actually allow a lot of code to just work.
-
     static partial class StringNormalizationExtensions
     {
         [SecurityCritical]
         public static bool IsNormalized(this string value, NormalizationForm normalizationForm)
         {
-            for (int i = 0; i < value.Length; i++)
+            if (value == null)
             {
-                if (value[i] > 0x7F)
-                {
-                    throw NotImplemented.ByDesign;
-                }
+                throw new ArgumentNullException("value");
+            }
+
+            if (normalizationForm != NormalizationForm.FormC && normalizationForm != NormalizationForm.FormD &&
+                normalizationForm != NormalizationForm.FormKC && normalizationForm != NormalizationForm.FormKD)
+            {
+                throw new ArgumentException(SR.Argument_InvalidNormalizationForm, "normalizationForm");
             }
 
-            return true;
+            if (HasInvalidUnicodeSequence(value))
+            {
+                throw new ArgumentException(SR.Argument_InvalidCharSequenceNoIndex, "value");
+            }
```

Comment:
I can factor out and share the null check as well as the normalizationForm argument checks.  The HasInvalidUnicodeSequenceCheck is specific to Unix as the Windows native API does it as a side effect of normalization, which differs from ICU.